File: example/Storage/astype.py
import sys
from pathlib import Path
home = str(Path.home())
sys.path.append(home + '/Cytnx_lib')
from cytnx import *


A = Storage(10)
print(A.dtype_str())

B = A.astype(Type.Double)
print(B.dtype_str())
print(B is A)

C = A.astype(Type.Float)
print(C.dtype_str())
print(C is A)

D = Storage(10,device=Device.cuda+0);
E = D.astype(Type.Float)
print(E.device_str())

File: example/Storage/astype.cpp
#include "cytnx.hpp"
#include <iostream>

using namespace cytnx;
int main() {
  /*
      1. Create a Storage with
      dtype =Type.Double [default],
  */
  Storage A(10);
  std::cout << A.dtype_str() << std::endl;

  // the type is the same as A's type, so B shares the same object with A
  Storage B = A.astype(Type.Double);
  std::cout << B.dtype_str() << std::endl;
  std::cout << is(B, A) << std::endl;  // true

  // cast A from Type.Double to Type.Float
  Storage C = A.astype(Type.Float);
  std::cout << C.dtype_str() << std::endl;
  std::cout << is(C, A) << std::endl;  // false

  Storage D(10, Type.Double, Device.cuda + 0);
  // D is on GPU, so E is also on GPU
  Storage E = D.astype(Type.Float);

  std::cout << E.device_str() << std::endl;

  return 0;
}
